lib/common/tests/strings/pcmk_parse_double.c

```diff
@@ -0,0 +1,135 @@
+#include <float.h>  // DBL_MAX, etc.
+#include <math.h>   // fabs()
+
+#include <glib.h>
+
+#include <crm_internal.h>
+
+#define PCMK__LOCAL_BUF_SIZE 2 * DBL_MAX_10_EXP
+
+/*
+ * Avoids compiler warnings for floating-point equality checks.
+ * Use for comparing numbers (e.g., 1.0 == 1.0), not expression values.
+ */
+#define PCMK__ASSERT_DBL_EQ(d1, d2) g_assert_cmpfloat(fabs(d1 - d2), \
+                                                      <, DBL_EPSILON);
+
+static void
+empty_input_string(void)
+{
+    // Without default_text
+    PCMK__ASSERT_DBL_EQ(pcmk_parse_double(NULL, NULL), PCMK__PARSE_DBL_DEFAULT);
+    g_assert_cmpint(errno, ==, EINVAL);
+
+    PCMK__ASSERT_DBL_EQ(pcmk_parse_double("", NULL), PCMK__PARSE_DBL_DEFAULT);
+    g_assert_cmpint(errno, ==, EINVAL);
+
+    // With default_text
+    PCMK__ASSERT_DBL_EQ(pcmk_parse_double(NULL, "2.0"), 2.0);
+
+    PCMK__ASSERT_DBL_EQ(pcmk_parse_double("", "2.0"), PCMK__PARSE_DBL_DEFAULT);
```

Comment:
I meant to ask you about that, actually. The above mirrors the behavior of crm_parse_ll() -- it uses the default text if text is NULL, but returns -1 if text is an empty string.

It makes sense from the perspective of how it handles string parsing in general. If you give scan_ll() (or scan_double()) a string that starts with non-numeric characters so that it fails to parse the string, it returns -1. In that sense, the empty string is just a bad string that it can't parse.

But it might be more intuitive to handle an empty string in the same way that we handle a NULL string. That's what we do in most places and what we do with pcmk__str_empty().

With that said, I don't know if we can change scan_ll()/crm_parse_ll()'s empty-string behavior since crm_parse_ll() is public. And for consistency, I would expect pcmk_parse_double() to handle an empty string in the same way as crm_parse_ll().

Reply:
Yep, that's another situation backward compatibility often puts us in. I think the main factor should be: how would an empty string arrive at this function, and if that involves the user at any point, how would they expect it to behave? If an empty string argument would never involve the user, then it doesn't matter, and we can just do something reasonable on the caller end. Also, since it's private API, it's not terribly important to decide now. But when we get around to making the legacy function names wrappers for new functions, we could change the behavior in the new function and hack it back to the old way in the wrapper.